components/nexus-repository/src/main/java/org/sonatype/nexus/repository/search/SearchFacetImpl.java

```diff
@@ -74,6 +75,34 @@ public SearchFacetImpl(final SearchService searchService,
   {
     this.searchService = checkNotNull(searchService);
     this.componentMetadataProducers = checkNotNull(componentMetadataProducers);
+    this.rebuildIndexLock = new ReentrantReadWriteLock();
+  }
+
+  @Override
+  @Guarded(by = STARTED)
+  public void rebuildIndex() {
+    log.info("Rebuilding index of repository {}", getRepository().getName());
+    rebuildIndexLock.writeLock().lock();
+    try {
+      searchService.deleteIndex(getRepository());
+      searchService.createIndex(getRepository());
+
+      // TODO: sort out with Stuart, is this correct, or would intermingle if this method is called from TX enabled method
+      try (final StorageTx tx = getRepository().facet(StorageFacet.class).txSupplier().get()) {
+        tx.begin();
+        for (Component component : tx.browseComponents(tx.getBucket())) {
+          try {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)